### DISCUSSION TITLE
Trigger actions on condition change

### DISCUSSION THREAD

**Discussion by b02860de585071a2** [slskd/slskd#922] [General]
Let's say that I have slskd running in the background on my home server, and would like to trigger a script to run when a download finishes.

Currently, the best way I can think to do this would be to schedule a cron job (or systemd timer) which runs something like:

```
# trigger [action] if [downloads folder] is not empty
if [[ ! -z "$(ls -A /path/to/slskd/downloads)" ]];
then
    [action];
else
    [other action]
fi
```

This is fairly clunky for a variety of reasons (cron job is only time-based and may trigger when a download isn't finished, etc), and I'm not a fan of it. 

Is there a straightforward way to:

- natively have the daemon execute arbitrary commands?
- read its state (externally) to act on?
- something else?

I appreciate any input here.

**Comment by jpdillingham** [comment]
I have a whole milestone for this ([Post-Download Automation](https://github.com/slskd/slskd/milestone/4)) but it is currently blocked on a number of other things.

The more I think about it, though, I'm thinking it won't harm to add some of this functionality now.  What's been blocking me on this is the need for a way to durably store some of these actions to facilitate retries, and things like sequencing issues.  This gets complicated if I want to, for example, run a number of different commands and then finally delete the downloaded file from disk; I don't want to delete it until everything else has run successfully.

Files are never deleted automatically today, so that shouldn't really block any sort of automation.

I think what I can offer is a way for users to execute a shell command and provide a URL to receive a webhook whenever a download finishes.  I'll have to figure out which parameters to pass in either case, but I think at a minimum the location of the file on disk is needed.

**Reply by b02860de585071a2** [reply]
I just glanced over that milestone:

> moving files between network shares, setting file permissions, post-processing metadata with a program like beets, or notifying external applications.

Wow, those are all things I'm looking to do. We're definitely on the same page there for use cases. 

> I don't want to delete it until everything else has run successfully.

Exactly the issue I've been running into, when trying to hack together a workaround.

I figured it would complicate things if you have operations that are still in progress (both in terms of downloads, and external programs which may not report their status) short of implementing some form of IPC - which is completely beyond the scope of this project. 

> a way for users to execute a shell command and provide a URL to receive a webhook whenever a download finishes

That sounds like a good option. Since it's isolated from the physical host machine, it wouldn't (shouldn't?) interfere with existing instances, docker containers, etc. 

It's been years since I've had to work with any C#, but from looking through the source code it appears that you have a pretty healthy webUI to build against here. 

If you need testers, I'm happy to get in line.
